### DISCUSSION TITLE
Is it possible to create a "handler" recipe?

### DISCUSSION THREAD

**Discussion by antonalekseev** [casey/just#1179] [Q&A] [answered]
Given the following `justfile`

```
 set dotenv-load := false

mkenv:
    @echo "Create env"
    echo "FOO=1" > .env

build-ok: mkenv
    @echo "Build status OK"
    #!/usr/bin/env sh
    cat .env
    exit 0

build-failed: mkenv
    #!/usr/bin/env sh
    cat .env
    exit 1

cleanup:
    rm .env
```

I wonder if there is any way to run the `cleanup` recipe automatically on error to handle `build-failed` or when all the recipes are done implicitly? Something like `trap` or https://docs.python.org/3/library/atexit.html.

**Comment by casey** [comment, marked as answer]
There isn't currently a way to do this. I've been thinking about it, but not sure how to do it in a reasonable way. One idea, which I don't think is particularly good, is to add a new kind of dependency:

```
foo:

bar: || foo
```

Dependencies after only `||` run if the recipe fails, analogously to the current `&& dependency` syntax, which runs after a recipe if it succeeds. I feel like this is probably too brittle, and won't work for enough use-cases to be worth it.

**Reply by patmaddox** [reply]
That is the exact syntax I thought of when I saw the `&&` syntax.

Our use case is we have what we call a "doctor" script, which checks for certain conditions on the machine (e.g. env files in place, certain binaries, whatever). If a check fails, we print a proposed remedy out to the screen.

I'll experiment more with just as-is, but I did think `||` would be helpful for our use case.

**Reply by casey** [reply]
Thanks for commenting! I'd definitely be open to implementing `||`, I have a weakness for weird features and weird syntax. Would you mind creating an issue describing your use-case, along with what behavior you'd like `||` to have? An example justfile would be great.

**Reply by jpeeler** [reply]
What if instead the subsequent dependency is prefixed with - to indicate to run always, especially in the case of error?

```
foo:
 
bar: && -foo
```

**Reply by casey** [reply]
I personally prefer `||`, it just seems more natural to me.
